src/evm.rs

```diff
@@ -516,19 +517,39 @@ impl<'a, Ext, Db: Database + DatabaseCommit, TrevmState: HasCfg> Trevm<'a, Ext,
     /// Run a function with the provided configuration, then restore the
     /// previous configuration. This will not affect the block and tx, if those
     /// have been filled.
-    pub fn with_cfg<F, C, NewState>(mut self, f: F, cfg: &C) -> Trevm<'a, Ext, Db, NewState>
+    pub fn with_cfg<C, F, NewState>(mut self, cfg: &C, f: F) -> Trevm<'a, Ext, Db, NewState>
     where
-        F: FnOnce(Self) -> Trevm<'a, Ext, Db, NewState>,
         C: Cfg,
+        F: FnOnce(Self) -> Trevm<'a, Ext, Db, NewState>,
         NewState: HasCfg,
     {
-        let previous = std::mem::take(self.inner.cfg_mut());
+        let previous = self.inner.cfg_mut().clone();
         cfg.fill_cfg_env(self.inner.cfg_mut());
         let mut this = f(self);
         *this.inner.cfg_mut() = previous;
         this
     }
 
+    /// Run a fallible function with the provided configuration, then restore the
+    /// previous configuration. This will not affect the block and tx, if those
+    /// have been filled.
+    pub fn try_with_cfg<C, F, NewState, E>(
+        mut self,
+        cfg: &C,
+        f: F,
+    ) -> Result<Trevm<'a, Ext, Db, NewState>, E>
+    where
+        C: Cfg,
+        F: FnOnce(Self) -> Result<Trevm<'a, Ext, Db, NewState>, E>,
+        NewState: HasCfg,
+    {
+        let previous = self.inner.cfg_mut().clone();
+        cfg.fill_cfg_env(self.inner.cfg_mut());
+        let mut this = f(self)?;
```

Comment:
this will not restore the cfg on failure, needs to follow pattern in `try_with_block`